nixos/modules/services/networking/teleport.nix

```
@@ -0,0 +1,96 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  cfg = config.services.teleport;
+  settingsYaml = pkgs.formats.yaml { };
+  pidFilePath = "/run/teleport.pid";
+in
+{
+  options = {
+    services.teleport = with lib.types; {
+      enable = mkEnableOption "the Teleport service";
+
+      settings = mkOption {
+        type = settingsYaml.type;
+        default = { };
+        example = literalExpression ''
+          {
+            teleport = {
+              nodename = "client";
+              advertise_ip = "192.168.1.2";
+              auth_token = "60bdc117-8ff4-478d-95e4-9914597847eb";
+              auth_servers = [ "192.168.1.1:3025" ];
+          }
+        '';
+        description = ''
+          Contents of the <literal>teleport.yaml</literal> config file.
+          The <literal>--config</literal> arguments will only be passed if this set is not empty.
+
+          See <link xlink:href="https://goteleport.com/docs/setup/reference/config/"/>.
+        '';
+      };
+
+      debug.enable = mkEnableOption "debug";
+
+      diag = {
+        enable = mkEnableOption ''
+          endpoints for monitoring purposes.
+
+          See <link xlink:href="https://goteleport.com/docs/setup/admin/troubleshooting/#troubleshooting/"/>
+        '';
+
+        addr = mkOption {
+          type = str;
+          default = "127.0.0.1";
+          description = "Metrics and diagnostics address.";
+        };
+
+        port = mkOption {
+          type = int;
+          default = 3000;
+          description = "Metrics and diagnostics port.";
+        };
+      };
+
+      extraOpts = mkOption {
+        type = str;
+        default = "";
+        example = "--labels=Foo=Bar --insecure";
+        description = ''
+          Extra commandline options.
+
+          See <link xlink:href="https://goteleport.com/docs/setup/reference/cli/#teleport-start"/>
+        '';
+      };
+
+    };
+  };
+
+  config = mkIf config.services.teleport.enable {
+    environment.systemPackages = [ pkgs.teleport ];
+
+    systemd.services.teleport = {
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+      serviceConfig = {
+        ExecStart = ''${pkgs.teleport}/bin/teleport start \
+          --pid-file=${pidFilePath} \
+          ${optionalString cfg.debug.enable "--debug"} \
+          ${optionalString cfg.diag.enable "--diag-addr=${cfg.diag.addr}:${toString cfg.diag.port}"} \
+          ${optionalString (cfg.settings != { }) "--config=${settingsYaml.generate "teleport.yaml" cfg.settings}"} \
+          ${cfg.extraOpts}
+        '';
```

minor nitpick: this would be more readable as
```suggestion
        ExecStart = ''
          ${pkgs.teleport}/bin/teleport start \
            --pid-file=${pidFilePath} \
            ${optionalString cfg.debug.enable "--debug"} \
            ${optionalString cfg.diag.enable "--diag-addr=${cfg.diag.addr}:${toString cfg.diag.port}"} \
            ${optionalString (cfg.settings != { }) "--config=${settingsYaml.generate "teleport.yaml" cfg.settings}"} \
            ${cfg.extraOpts}
        '';
```
which would also insert fewer spaces into the generated directive. (an empty initial line in `''`-quoted strings is ignored completely)

less minor nitpick: this still subjects `extraOpts` to systemd command line expansion. if it's likely that `extraOpts` will never contain a `%`-specifier or `${name}` substrings we'll stop harking about that, but if those are likely in arguments this does need some escaping to not be surprising to users. (this also applies to lots of other services currently in nixpkgs and should be fixed treewide eventually)

otherwise LGTM